### PR TITLE
Feature | Parse time

### PR DESCRIPTION
**Pull request by h-dong**
Now the library will detect the following time formats:
- at 8:36am
- by 12pm
- 8:30am
- 10:50pm
- 01:08am
- 09:45
- 9am
- 7pm
- 12pm
- 12am
- 0:00
- 11:00

After refactor, now also support:
- 12 am (with space)
- 12 a.m. or 12a.m. (with periods)
- 12AM (capitals)

### CODE REVIEW

File: src/filters/time/time.constants.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+const TIME = {
+    ALL: '((([1-9]|(0|1)[0-2])(:[0-5][0-9])?(a|p)m)\\b|\\b((((0|1)?[0-9])|(2[0-3]))(:[0-5][0-9])))',
```
Comment (MatthiasGu) [inline src/filters/time/time.constants.ts]: This regex will not catch the following:
- 12 am (with space)
- 8 or 15 or any single number (does accept 8:00)
- 12 a.m. or 12a.m. (with periods)
- 12AM (capitals)
- noon or midnight (words frequently used to replace 12am or 12pm)


I will leave it up to you to decide about the final 2 points, but I think the first 3 should certainly be fixed.

File: src/filters/time/time.test.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            case: 'have lunch 12pm',
+            result: [
+                { dateTime: mockTime(12, 0), text: 'have lunch', matched: '12pm' },
+            ],
+        },
+        {
+            case: 'get ready for bed 12am',
+            result: [
+                { dateTime: mockTime(0 ,0), text: 'get ready for bed', matched: '12am' },
```
Comment (MatthiasGu) [inline src/filters/time/time.test.ts]: Super pedantic, but space in mocktime(0, 0) is wrong, should be afterComma 🗡

File: src/filters/time/time.test.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import { TestCaseSchema } from '../../serina.schema';
+import Time from './time';
+
+describe('Time', () => {
+    const mockTime = (hour, minute) => DateTime.utc()
+        .set({ hour, minute })
+        .startOf('minute')
+        .toJSDate();
+
+    const testCases: TestCaseSchema[] = [
```
Comment (MatthiasGu) [inline src/filters/time/time.test.ts:11]: I think it is possible to do it with loops and test at least all possible hours. Might seem like an overkill, but since the regex is quite exact, it might be good to have more tests. What do you think?

File: src/filters/time/time.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            if (contains(matchingText, timePattern)) {
+                // match again without filler words
+                // then take the first and only matched string
+                const time = matchPattern(matchingText, TIME.ALL)[0];
+
+                let replaceAm;
+                let replacePm;
+                let timeSplit;
+
+                switch (timePattern) {
```
Comment (MatthiasGu) [inline src/filters/time/time.ts]: I think we can come up with a better algorithm for this and avoid the switch, but it is not that simple. Perhaps could be a tech task for the future.

**Review by MatthiasGu — CHANGES_REQUESTED** [review]
Need to catch a couple more cases with your regex. Other than that, looks good!

File: src/filters/time/time.ts (hunk 4 above)
Comment (h-dong) [reply src/filters/time/time.ts]: Yeah, I'd say leave this for later. Abstracting it into a some function will be easier to read. But I'd much prefer to do that once we have done the relative time, as this might all get changed again when that comes in.

File: src/filters/time/time.test.ts (hunk 2 above)
Comment (h-dong) [reply src/filters/time/time.test.ts]: Fixed

File: src/filters/time/time.test.ts (hunk 3 above)
Comment (h-dong) [reply src/filters/time/time.test.ts:11]: I think it is a bit of an overkill, because the regex is quite straightforward. Testing 8:30 vs 9:30 adds very little value to me, if you feel we could cover more edge cases then I'd be fine with that. I don't think looping every hour is worth it, especially then the questions becomes should we loop all the minutes too?

My philosophy is add enough test to cover the key edge cases, every time you find a bug. Fix it then add a test to prevent similar bugs coming up in the future.

File: src/filters/time/time.constants.ts (hunk 1 above)
Comment (h-dong) [reply src/filters/time/time.constants.ts]: What do you mean by "8 or 15 or any single number (does accept 8:00)"?

Comment (h-dong) [reply src/filters/time/time.constants.ts]: - Capitals I'll look at now, that should be handled
- in terms of space, full-stops, I'll see how much effort it is to add in
- noon or midnight should be done as part of relative time I think

File: src/filters/time/time.constants.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+const MERIDIEM = '(\\.)?m(\\.)?';
+const ANTE_MERIDIEM = `(a${MERIDIEM})`;
+const POST_MERIDIEM = `(p${MERIDIEM})`;
+const ALL_MERIDIEM = `((a|p)${MERIDIEM})`;
+
+const TIME = {
+    ALL: `[0-9]{1,2}(((:[0-5][0-9])(( )?${ALL_MERIDIEM}))|(:[0-5][0-9])|(( )?${ALL_MERIDIEM}))`,
```
Comment (MatthiasGu) [inline src/filters/time/time.constants.ts:7]: Still doesn't match single numbers, e.g. if you write "Go to dentist at 8" it won't match but it will match "Go to dentist at 8:00"

File: src/filters/time/time.test.ts (hunk 3 above)
Comment (MatthiasGu) [reply src/filters/time/time.test.ts:11]: Fair enough, I just thought it is nice to automate as much as possible.

File: src/utils/convertTimeStringToObj.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import { TimeObjectSchema } from 'serina.schema';
+import { POST_MERIDIEM, ALL_MERIDIEM } from 'filters/time/time.constants';
+import contains from './contains';
+import { remove, isValidTime } from 'utils';
+
+function convertTimeStringToObj(time: string): TimeObjectSchema {
```
Comment (MatthiasGu) [inline src/utils/convertTimeStringToObj.ts:6]: gooda jobba

File: src/utils/trimWhiteSpaces.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+function trimWhiteSpaces(text: string): string {
+    const trimedText = text.replace(/  /g, ' ');
```
Comment (MatthiasGu) [inline src/utils/trimWhiteSpaces.ts]: Minor point, trimmed should have double "m" :eye:

File: src/utils/trimWhiteSpaces.test.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+    test('should not trip anything extra', () => {
+        const check = trimWhiteSpaces(text);
+        expect(check).toBe('this is my string');
+    });
+
+    test('should trip white space in between the string', () => {
+        text = 'this is my  string';
+        const check = trimWhiteSpaces(text);
+        expect(check).toBe('this is my string');
```
Comment (MatthiasGu) [inline src/utils/trimWhiteSpaces.test.ts:18]: Since you are setting text in beforeEach, you could check against that (call it expectedText or something).

File: src/utils/remove.test.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    beforeEach(() => {
+        text = 'this is my test string';
+        pattern = 'test';
+    });
+
+    test('should remove pattern from text', () => {
+        const check = remove(text, pattern);
+        expect(check).toBe('this is my  string');
+    });
+
```
Comment (MatthiasGu) [inline src/utils/remove.test.ts:16]: I would think that it shouldn't create an extra space. I know why it does it, but don't you think it's an unwanted side effect?

File: src/utils/isValidTime.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+function isValidTime(hour, minute) {
+    if (hour < 0 || hour > 23 || minute < 0 || minute > 59) return false;
+
+    return true;
+}
```
Comment (MatthiasGu) [inline src/utils/isValidTime.ts:3]: Seems like a good use case for ternary operator :Java:

File: src/utils/isValidTime.test.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    test('should not allow 24:00', () => {
+        expect(isValidTime(24, 0)).toBe(false);
+    });
+
+    test('should not allow 25:00', () => {
+        expect(isValidTime(25, 0)).toBe(false);
+    });
+
+    test('should not allow 23:60', () => {
+        expect(isValidTime(25, 0)).toBe(false);
```
Comment (MatthiasGu) [inline src/utils/isValidTime.test.ts]: That's not what the test is doing ;)

File: src/utils/convertTimeStringToObj.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    let minute;
+
+    if (contains(time, `( )?${ALL_MERIDIEM}`, false)) {
+        // 12H
+        const timeWithMeridiem = remove(time, `( )?${ALL_MERIDIEM}`, false);
+        const timeSplit = timeWithMeridiem.split(':');
+        hour = parseInt(timeSplit[0], 10);
+        minute = parseInt(timeSplit[1], 10);
+
+        if (isNaN(minute)) minute = 0;
```
Comment (MatthiasGu) [inline src/utils/convertTimeStringToObj.ts:17]: I think if we are going to allow whole hours to be written as "8" rather than "8:00" the above logic might change slightly. The split on (':') is still fine, but then we need to check for minute to be undefined.

File: src/utils/convertTimeStringToObj.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+        if (contains(time, POST_MERIDIEM, false)) {
+            // Post Meridiem (PM)
+            if (hour < 12) hour += 12;
+        } else {
+            // Ante Meridiem (AM)
+            if (hour === 12) hour = 0;
+        }
+    } else {
+        // 24H
```
Comment (MatthiasGu) [inline src/utils/convertTimeStringToObj.ts:27]: Ditto.

**Review by MatthiasGu — CHANGES_REQUESTED** [review]
A few comments and small changes...

File: src/filters/time/time.constants.ts (hunk 5 above)
Comment (h-dong) [reply src/filters/time/time.constants.ts:7]: We discussed this already

File: src/filters/time/time.test.ts (hunk 3 above)
Comment (h-dong) [reply src/filters/time/time.test.ts:11]: Aye, maybe later. For now I think these tests should be enough

File: src/utils/convertTimeStringToObj.ts (hunk 6 above)
Comment (h-dong) [reply src/utils/convertTimeStringToObj.ts:6]: 🥇

File: src/utils/convertTimeStringToObj.ts (hunk 12 above)
Comment (h-dong) [reply src/utils/convertTimeStringToObj.ts:17]: See without doing that, this logic is simpler too lol

File: src/utils/isValidTime.test.ts (hunk 11 above)
Comment (h-dong) [reply src/utils/isValidTime.test.ts]: Lol, forgot to change it. Will update it

File: src/utils/isValidTime.ts (hunk 10 above)
Comment (h-dong) [reply src/utils/isValidTime.ts:3]: Good idea

File: src/utils/remove.test.ts (hunk 9 above)
Comment (h-dong) [reply src/utils/remove.test.ts:16]: I actually left it like that, but I see your point. I've changed it now, double white space removed.

File: src/utils/trimWhiteSpaces.test.ts (hunk 8 above)
Comment (h-dong) [reply src/utils/trimWhiteSpaces.test.ts:18]: I decided to overwrite it instead, thus why it was declared as `let` instead of `const`. I see your point, but I think I'm gonna leave it like this for now because I followed this technique in more places. Maybe in the future I'll change all of them at once.

File: src/utils/trimWhiteSpaces.ts (hunk 7 above)
Comment (h-dong) [reply src/utils/trimWhiteSpaces.ts]: Fixed

File: src/utils/isValidTime.ts (hunk 10 above)
Comment (MatthiasGu) [reply src/utils/isValidTime.ts:3]: Lol now that I look at this, you don't need ternary at all, you can just do a return since it's a boolean 🤣

Comment (h-dong) [reply src/utils/isValidTime.ts:3]: I'm gonna stick with this, otherwise it'll be too hard to read with `return !()`
